life_expectancy/data_loading.py

```diff
@@ -1,15 +1,107 @@
+    @abstractmethod
+    def load_data(self, filepath: str) -> pd.DataFrame:
+        """
+        Abstract method for loading the data.
+        """
 
-def save_data(dataframe: pd.DataFrame, filepath: str) -> None:
+    @abstractmethod
+    def normalize_data(self, dataframe: pd.DataFrame) -> pd.DataFrame:
+        """
+        Abstract method for normalizing the data.
+        """
+
+class TsvLoader(DataLoader):
+    """
+    Concrete class for TSV data loading.
+    """
+
+    def load_data(self, filepath: str) -> pd.DataFrame:
+        """
+        Function responsible for loading the data.
+        """
+        dataframe = pd.read_csv(filepath, delimiter='\t')
+
+        return dataframe
+
+    def normalize_data(self, dataframe: pd.DataFrame) -> pd.DataFrame:
+        """
+        Function responsible for normalizing the dataframe.
+        """
+        # Split the first column into separate columns
+        dataframe[['unit', 'sex', 'age', 'region']] = dataframe['unit,sex,age,geo\\time']\
+            .str.split(',', expand=True)
+
+        # Drop the original composed column
+        dataframe = dataframe.drop(columns=['unit,sex,age,geo\\time'])
+
+        # Unpivot data to long format
+        dataframe = dataframe.melt(id_vars=['unit','sex','age','region'], \
+            var_name='year', value_name='value')
+
+        return dataframe
+
+class JsonLoader(DataLoader):
+    """
+    Concrete class for JSON data loading.
+    """
+
+    def load_data(self, filepath: str) -> pd.DataFrame:
+        """
+        Function responsible for loading the data.
+        """
+        dataframe = pd.read_json(filepath)
+
+        return dataframe
+
+    def normalize_data(self, dataframe: pd.DataFrame) -> pd.DataFrame:
+        # Drop columns "flag" and "flag_detail"
+        dataframe = dataframe.drop(["flag", "flag_detail"], axis=1)
+
+        # Rename columns "country" and "life_expectancy"
+        dataframe = dataframe.rename(columns={"country": "region", "life_expectancy": "value"})
+
+        # Select only the columns that we need
+        dataframe = dataframe[["unit", "sex", "age", "region", "year", "value"]]
+
+        return dataframe
+
+class LoaderPicker:
     """
-    Function responsible for saving the data to a file.
+    Class responsible for executing the loader based on context.
     """
-    dataframe.to_csv(filepath, index=False)
+
+    def __init__(self, dataloader: DataLoader) -> None:
+        self.dataloader = dataloader
+
+    def load_df(self, filepath: str) -> pd.DataFrame:
+        """
+        Function responsible for calling load_data method
+        """
+        dataframe = self.dataloader.load_data(self,filepath)
+        return dataframe
+
+    def normalize_df(self, dataframe: pd.DataFrame) -> pd.DataFrame:
+        """
+        Function responsible for calling normalize_data method
+        """
+        dataframe = self.dataloader.normalize_data(self,dataframe)
+        return dataframe
+
+class DataSaver:
```

Comment:
sugestão: pq não criar uma classe que faz o file_handling com os métodos load e save, para gerir operações de ficheiros de forma mais centralizada?

Reply:
Confesso que não percebi esta sugestão, podes dar um exemplo?

Reply:
claro!

class FileHandler:
    """Class to load and save files."""

    def __init__(self, strategy=None):
        self.strategy = strategy

    def save_data(
        self, df_final: pd.DataFrame, output_file_path: Path, region_filter: str
    ) -> None:
        """Save the final dataframe to a file."""
        if df_final is None:
            logging.warning("The final dataframe is None. Nothing will be saved.")
            return

        try:
            df_final.to_csv(output_file_path, index=False)
        except PermissionError:
            logging.error(
                "Output file %s could not be created or written to.", output_file_path
            )
            return

        logging.info(
            "Successfully saved cleaned data for region %s at %s",
            region_filter,
            output_file_path,
        )

    def load_data(self, input_file_path: Union[str, Path]) -> pd.DataFrame:
        """Load the data from a file."""
        if self.strategy:
            return self.strategy.load_data(input_file_path)
        logging.error("No file handling strategy has been set.")
        return pd.DataFrame()